### PR TITLE
Add Sunrise / Sunset Times Extension

### PR DESCRIPTION
**Pull request by jeffreydyao**
## Description

Check sunrise / sunset times for dates up to a month in the future at your current location.

First Raycast extension - whoo!

**Commands:**
- Check Sunrise / Sunset Times: See sunrise / sunset times for dates up to a month in the future.

**Actions:**
- Copy Sunrise Time: Copy the currently selected date's sunrise time to the clipboard (i.e. _April 7 | Sunrise: 6:12:26 AM_)
- Copy Sunset Time: Copy the currently selected date's sunset time to the clipboard (i.e. _April 7 | Sunset: 5:44:45 PM_)

sunrise-sunset-times-01
sunrise-sunset-times-02
sunrise-sunset-times-03

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

File: extensions/sunrise-sunset-times/src/index.tsx [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    >
+      {dateArray.map((date, i) => (
+        <List.Item
+          key={i}
+          title={dayjs(date).format("MMMM D")}
+          subtitle={
+            timesLoading
+              ? "Calculating ..."
+              : "☀ " + dayjs(sunrise[i]).format("h:mm:ss A") + "   ☾ " + dayjs(sunset[i]).format("h:mm:ss A")
+          }
```
Comment (pernielsentikaer) [inline extensions/sunrise-sunset-times/src/index.tsx:106]:
```suggestion
            accessories={[
              { icon: "☀️", text: dayjs(sunrise[i]).format("h:mm A") },
              { icon: "🌜", text: dayjs(sunrise[i]).format("h:mm A") },
            ]}
```

File: extensions/sunrise-sunset-times/src/index.tsx [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  const [timesLoading, setTimesLoading] = useState(true);
+
+  // TODO: Type API response
+  /*   interface ILocation {
+    lat: number;
+    lon: number;
+    city: string;
+    country: string;
+  } */
+
```
Comment (pernielsentikaer) [inline extensions/sunrise-sunset-times/src/index.tsx:30]: Leftover?

File: extensions/sunrise-sunset-times/src/index.tsx [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  const sunriseArray: Date[] = [];
+  const sunsetArray: Date[] = [];
+
+  const [city, setCity] = useState();
+  const [country, setCountry] = useState();
+  const [sunrise, setSunrise] = useState<Date[]>([]);
+  const [sunset, setSunset] = useState<Date[]>([]);
+  // Hooks to toggle loading states
+  const [locationLoading, setLocationLoading] = useState(true);
+  const [timesLoading, setTimesLoading] = useState(true);
```
Comment (pernielsentikaer) [inline extensions/sunrise-sunset-times/src/index.tsx:21]: Seems like it's not in use.

File: extensions/sunrise-sunset-times/src/index.tsx [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      const times = suncalc.getTimes(date, lat, lng);
+      // Push sunrise/sunset times for current date to respective arrays.
+      sunriseArray.push(times.sunrise);
+      sunsetArray.push(times.sunset);
+    });
+
+    setSunrise(sunriseArray);
+    setSunset(sunsetArray);
+
+    setTimesLoading(false);
```
Comment (pernielsentikaer) [inline extensions/sunrise-sunset-times/src/index.tsx:79]: Is this being used?

File: extensions/sunrise-sunset-times/src/index.tsx [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        <List.Item
+          key={i}
+          title={dayjs(date).format("MMMM D")}
+          subtitle={
+            timesLoading
+              ? "Calculating ..."
+              : "☀ " + dayjs(sunrise[i]).format("h:mm:ss A") + "   ☾ " + dayjs(sunset[i]).format("h:mm:ss A")
+          }
+          actions={
+            <ActionPanel title="Copy times">
```
Comment (pernielsentikaer) [inline extensions/sunrise-sunset-times/src/index.tsx:108]:
```suggestion
            <ActionPanel title="Copy Times">
```

File: extensions/sunrise-sunset-times/src/index.tsx [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+              <Action.CopyToClipboard
+                title="Copy Sunrise Time"
+                content={dayjs(date).format("MMMM D") + " | Sunrise: " + dayjs(sunrise[i]).format("h:mm:ss A")}
```
Comment (pernielsentikaer) [inline extensions/sunrise-sunset-times/src/index.tsx:111]:
```suggestion
                content={dayjs(date).format("MMMM D") + " | Sunrise: " + dayjs(sunrise[i]).format("h:mm A")}
```
Removed seconds

File: extensions/sunrise-sunset-times/src/index.tsx [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+              />
+              <Action.CopyToClipboard
+                title="Copy Sunset Time"
+                content={dayjs(date).format("MMMM D") + " | Sunset: " + dayjs(sunset[i]).format("h:mm:ss A")}
```
Comment (pernielsentikaer) [inline extensions/sunrise-sunset-times/src/index.tsx:115]:
```suggestion
                content={dayjs(date).format("MMMM D") + " | Sunset: " + dayjs(sunset[i]).format("h:mm A")}
```

File: extensions/sunrise-sunset-times/package.json [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  "categories": [
+    "Other",
+    "Fun"
+  ],
+  "license": "MIT",
+  "commands": [
+    {
+      "name": "index",
+      "title": "Check Sunrise / Sunset Times",
+      "subtitle": "Up to a month in the future",
```
Comment (pernielsentikaer) [inline extensions/sunrise-sunset-times/package.json:17]:
```suggestion
      "title": "Sunrise & Sunset Times",
```

File: extensions/sunrise-sunset-times/package.json [hunk 9]
```diff
@@ -0,0 +1,46 @@
+{
+  "$schema": "https://www.raycast.com/schemas/extension.json",
+  "name": "sunrise-sunset-times",
+  "title": "Sunrise / Sunset Times",
```
Comment (pernielsentikaer) [inline extensions/sunrise-sunset-times/package.json:4]:
```suggestion
  "title": "Sunrise & Sunset Times",
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @jeffreydyao 👋 

Congratulations on your first extension. Thanks a lot! 🔥 

I have some feedback and suggestions for you to address:

- You should use the [IsLoading property for your list](https://developers.raycast.com/api-reference/user-interface/list#props) and first provide content when all data is present, the flicker doesn't do anything good.

- If it's not possible to find the position, then you should return an empty array instead.

- Could you move the CHANGELOG.md to the root folder

Other ideas for later:

- You could also add the possibility to add "custom locations" and combine it with a [dropdown component](https://developers.raycast.com/api-reference/user-interface/list#list.dropdown)

Request a new review when you are ready. Feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by jeffreydyao** [comment]
> Hi @jeffreydyao 👋
> 
> Congratulations on your first extension. Thanks a lot! 🔥
> 
> I have some feedback and suggestions for you to address:
> 
> * You should use the [IsLoading property for your list](https://developers.raycast.com/api-reference/user-interface/list#props) and first provide content when all data is present, the flicker doesn't do anything good.
> * If it's not possible to find the position, then you should return an empty array instead.
> * Could you move the CHANGELOG.md to the root folder
> 
> Other ideas for later:
> 
> * You could also add the possibility to add "custom locations" and combine it with a [dropdown component](https://developers.raycast.com/api-reference/user-interface/list#list.dropdown)
> 
> Request a new review when you are ready. Feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

Hi, thanks for the feedback! I'll fix it up after my uni exams.

**Comment by pernielsentikaer** [comment]
@jeffreydyao - That sounds like a good plan. I'll draft it until you're ready 😄

**Comment by pernielsentikaer** [comment]
I'm closing this due to inactivity; feel free to open it again or comment in the thread when you're ready to continue working on it 🙂

You can also catch me in [Slack](https://www.raycast.com/community) if you want to discuss 😄

**Pull request closed without merging**